implementations/rust/ockam/ockam/src/system/handler.rs

```diff
@@ -0,0 +1,23 @@
+use ockam_core::{Address, Message, Result, Routed};
+
+/// Handle a single type of message for a worker system-address
+///
+/// A handle may re-emit messages to the worker system, or to the
+/// Ockam runtime.  All state associated with a particular protocol
+/// must be contained in the type that implements this trait.
+#[ockam_core::async_trait]
+pub trait SystemHandler<C, M>
+where
+    C: Send + 'static,
+    M: Message,
+{
+    /// Run an initialise hook
+    ///
+    /// This gives this SystemHandler the chance to setup any state on
+    /// the worker or send messages to other workers (and hooks).
+    /// Finally it must return the address it wants to respond to
+    async fn initialize(&mut self, ctx: &mut C) -> Result<Address>;
```

Comment:
We decide what is the correct address for a particular Worker when we start it, is there a reason why for SystemHandler we delegate that responsibility to the SystemHandler instance itself?

Reply:
The reason why I left this up to the SystemHandler instance at the moment is that potentially a worker could add new SystemHandlers during runtime in response to external commands. But I do also think that this makes initialisation easier (you just have to pass a SystemHandler istance, not that _and_ and address you can potentially typo.